btree/btree.c

```diff
@@ -3529,164 +3565,561 @@ int64_t btree_iter_tick(struct m0_sm_op *smop)
+			node_slot.s_node = oi->i_nop.no_node;
+			lev->l_seq = lev->l_node->n_seq;
+			/* verify node footer */
+			int rc = node_verify(lev->l_node);
+			if (rc)
+			{
+				node_op_fini(&oi->i_nop);
+				return fail(bop, rc);
+			}
+			oi->i_nop.no_node = NULL;
+
+			oi->i_key_found = node_find(&node_slot,
+						    &bop->bo_rec.r_key);
+			lev->l_idx = node_slot.s_idx;
+
+			if (node_level(node_slot.s_node) > 0) {
+				if (oi->i_key_found) {
+					lev->l_idx++;
+					node_slot.s_idx++;
+				}
+				node_child(&node_slot, &child_node_addr);
+
+				if (!address_in_segment(child_node_addr)) {
+					node_op_fini(&oi->i_nop);
+					return fail(bop, M0_ERR(-EFAULT));
+				}
+				oi->i_used++;
+				return node_get(&oi->i_nop, tree,
+						&child_node_addr, P_NEXTDOWN);
+			} else {
+				if (!oi->i_key_found)
+					return P_LOCK;
+				/**
+				 * If root is an internal node and it contains
+				 * only two record, if any of the record is
+				 * going to be deleted, load the other child of
+				 * root.
+				 */
+				if (oi->i_used > 0 &&
+				    node_count_rec(oi->i_level[0].l_node) == 2)
+					return root_case_handle(bop);
+
+				return P_LOCK;
+			}
+		} else {
+			node_op_fini(&oi->i_nop);
+			return fail(bop, oi->i_nop.no_op.o_sm.sm_rc);
+		}
+	case P_STORE_CHILD: {
+		/* store child of the root. */
+		oi->i_level[1].l_sibling = oi->i_nop.no_node;
+		oi->i_level[1].l_sib_seq = oi->i_nop.no_node->n_seq;
+		/* Fall through to the next step */
+	}
+	case P_LOCK:
+		if (!locked(tree))
+			return lock_op_init(&bop->bo_op, &bop->bo_i->i_nop,
+					    bop->bo_arbor->t_desc, P_CHECK);
+		else
+			return P_CHECK;
+	case P_CHECK: {
+		oi->i_trial++;
+		if (!path_check(oi, tree, &bop->bo_rec.r_key.k_cookie) ||
+		    !child_node_check(oi)) {
+			if (oi->i_trial == MAX_TRIALS) {
+				if (bop->bo_flags & BOF_LOCKALL)
+					return fail(bop, -ETOOMANYREFS);
+				else
+					bop->bo_flags |= BOF_LOCKALL;
+			}
+			if (bop->bo_arbor->t_height != tree->t_height) {
+				/* If height has changed */
+				lock_op_unlock(tree);
+				level_cleanup(oi, bop->bo_tx);
+				return P_INIT;
+			} else {
+				/* If height is same */
+				lock_op_unlock(tree);
+				return P_DOWN;
+			}
+		}
+		/* Fall through to the next step */
+	}
+	case P_ACT: {
+		struct m0_btree_rec rec;
+		struct slot         node_slot;
+		/**
+		 *  if key exists, delete the key, if there is an underflow, go
+		 *  to resolve function else return P_CLEANUP.
+		*/
+
+		if (!oi->i_key_found)
+			rec.r_flags = M0_BSC_KEY_NOT_FOUND;
+		else {
+			lev = &oi->i_level[oi->i_used];
+			node_slot.s_node = lev->l_node;
+			node_slot.s_idx  = lev->l_idx;
+			node_del(node_slot.s_node, node_slot.s_idx, bop->bo_tx);
+			lev->l_node->n_skip_rec_count_check = true;
+			node_done(&node_slot, bop->bo_tx, true);
+			node_fix(node_slot.s_node, bop->bo_tx);
+
+			rec.r_flags = M0_BSC_SUCCESS;
+		}
+		int rc = bop->bo_cb.c_act(&bop->bo_cb, &rec);
+		if (rc) {
+			lock_op_unlock(tree);
+			return fail(bop, rc);
+		}
```

Comment:
If the callback returns error then underflow will not be resolved at upper levels. Is this expected behavior ?

Reply:
As, expected error code will be KEY_NOT_FOUND, if there is no key , there will be no underflow

Reply:
The error code (rc) is from callback, it will not be KEY_NOT_FOUND.